### PR TITLE
New ReceiveTimeout not applied if it was set while handling a  messag…

### PR DESCRIPTION
**Pull request by helena**
## Purpose

Adds a notion of `beforeReceive` and captures receive timeout duration and task, updated on each `ReceiveTimeout.setReceiveTimeout` so that on evaluation of `ReceiveTimeout.checkReceiveTimeoutIfNeeded` can be checked for changes if modified while handling `NotInfluenceReceiveTimeout`.
 
## References
https://github.com/akka/akka/issues/26880
 
## Changes
* ReceiveTimeout handling if it was set while handling a `NotInfluenceReceiveTimeout` message
* Tests related to change - more exact on expectation of timeout on change

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala [hunk 1]
```diff
@@ -12,56 +12,65 @@ import scala.concurrent.duration.Duration
...
-  private var receiveTimeoutData: (Duration, Cancellable) = emptyReceiveTimeoutData
+  private var value: ReceiveTimeoutValue = ReceiveTimeoutValue()
 
-  final def receiveTimeout: Duration = receiveTimeoutData._1
+  final def receiveTimeout: Duration = value.current
 
-  final def setReceiveTimeout(timeout: Duration): Unit = receiveTimeoutData = receiveTimeoutData.copy(_1 = timeout)
+  final def setReceiveTimeout(timeout: Duration): Unit = {
+    val prev = value.current
+    value = value.copy(current = timeout, previous = prev)
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: Not sure that's the right place to track the previous entry. What happens if `setReceiveTimeout` is called multiple times during processing of a message? I think we something that is called before a message is processed and something afterwards and then compare those values.

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala [hunk 2]
```diff
@@ -12,56 +12,65 @@ import scala.concurrent.duration.Duration
...
+  }
 
   protected def checkReceiveTimeoutIfNeeded(message: Any): Unit =
     if (hasTimeoutData)
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: I think this still misses if a timeout was removed during a `NotInfluenceReceiveTimeout` message handling.

**Comment by helena** [comment]
Added MIMA filters

**Review by jrudolph — COMMENTED** [review]
Would probably be good to add a few new tests:
 * remove existing timeout while handling a `NotInfluenceReceiveTimeout` message
 * change existing timeout while handling a `NotInfluenceReceiveTimeout` message
 * while handling one such method, it should still work correctly if you change the timeout multiple times

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala (hunk 1 above)
Comment (helena) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: In my first WIP I had it saving previous in ActorCell.invoke before processing of the message occurs. Is that where you mean?

Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: yes, that was my thought, returning previous from `cancelReceiveTimeoutIfNeeded` which is before processing message

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala [hunk 3]
```diff
@@ -12,56 +12,65 @@ import scala.concurrent.duration.Duration
...
     }
-  }
 
   private def rescheduleReceiveTimeout(f: FiniteDuration): Unit = {
-    receiveTimeoutData._2.cancel() //Cancel any ongoing future
-    val task = system.scheduler.scheduleOnce(f, self, akka.actor.ReceiveTimeout)(this.dispatcher)
-    receiveTimeoutData = (f, task)
+    value.task.cancel() //Cancel any ongoing future
+    val t = system.scheduler.scheduleOnce(f, self, akka.actor.ReceiveTimeout)(this.dispatcher)
+    value = value.copy(current = f, task = t) // keep previous
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: Since we do this for each message (and also the cancel) when receive timeout is used it is worth optimizing this by using `var` fields in `ReceiveTimeoutValue` and mutate them instead of copy. Avoids allocations. Doesn't have to be in this PR if we think that is an unrelated change.

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala [hunk 4]
```diff
@@ -12,56 +12,65 @@ import scala.concurrent.duration.Duration
...
-  final def receiveTimeout: Duration = receiveTimeoutData._1
+  final def receiveTimeout: Duration = value.current
 
-  final def setReceiveTimeout(timeout: Duration): Unit = receiveTimeoutData = receiveTimeoutData.copy(_1 = timeout)
+  final def setReceiveTimeout(timeout: Duration): Unit = {
+    val prev = value.current
+    value = value.copy(current = timeout, previous = prev)
+  }
 
   protected def checkReceiveTimeoutIfNeeded(message: Any): Unit =
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: Can we have `prev` as a parameter to this method instead of having it inside ReceiveTimeoutValue? It's only used within one call of ActorCell and doesn't have to be stored longer.

Comment (helena) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: Yes, and I reverted the case class and mima.

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala (hunk 2 above)
Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: This comment is still valid, because if it's a NotInfluenceReceiveTimeout message it will not be cancelled.
Like this:
```
  val receiveTimeoutChanged = receiveTimeoutData ne beforeReceive
  if (hasTimeoutData || receiveTimeoutChanged)
    checkReceiveTimeout(!message.isInstanceOf[NotInfluenceReceiveTimeout] || receiveTimeoutChanged)
```

File: akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala [hunk 5]
```diff
@@ -179,15 +179,59 @@ class ReceiveTimeoutSpec extends AkkaSpec {
...
+
+    "remove / change existing timeout while handling a NotInfluenceReceiveTimeout message" taggedAs TimingTest in {
+      val timeoutLatch = TestLatch()
+      val initialTimeout = 500.millis
+
+      val timeoutActor = system.actorOf(Props(new Actor {
+        context.setReceiveTimeout(initialTimeout)
+
+        def receive: Receive = {
+          case Identify(_)    => context.setReceiveTimeout(Duration.Undefined)
```
[the remark this answers is not in the document]
Comment (helena) [reply akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: Thanks, forgot about that. I use `TransparentTick` farther down in the other new tests, wanted to try a different `NotInfluenceReceiveTimeout`.

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala (hunk 2 above)
Comment (helena) [reply akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala]: This one really schooled me :) Thanks for the input, pushed change.

File: akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala [hunk 6]
```diff
@@ -179,15 +179,59 @@ class ReceiveTimeoutSpec extends AkkaSpec {
...
+            context.setReceiveTimeout(initialTimeout)
+
+          case ReceiveTimeout => timeoutLatch.open
+        }
+      }))
+
+      timeoutActor ! TransparentTick
+      timeoutActor ! TransparentTick
+
+      Await.ready(timeoutLatch, TestLatch.DefaultTimeout)
```
Comment (jrudolph) [inline akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: Not quite clear to me how this test would fail? Did it fail before your changes?

Comment (helena) [reply akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: Are you saying it should fail? It does not fail and did not before.

Comment (patriknw) [reply akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: One way of testing it is to use a TestProbe instead of the latch. Send to the probe in `case ReceiveTimeout =>`.

then
```
// not triggered by initialTimeout because it was changed by the ticks
probe.expectNoMessage(initialTimeout + initialTimeout/2)
// now should happen
probe.expectMessage("ReceiveTimeout triggered")
```

Comment (helena) [reply akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: On second look this addition seems correct - on processing a `NotInfluenceReceiveTimeout`, if the timeout is modified multiple times, and here, given the longer timeout, it would not fail. However if we had it wait for a shorter duration, less than the last it would.

Comment (helena) [reply akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: This works if it first tests the last value used in `context.setReceiveTimeout` vs `initialTimeout + initialTimeout/2`. The probe receives the `ReceiveTimeout` within ~620 ms.

File: akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala [hunk 7]
```diff
@@ -179,15 +179,62 @@ class ReceiveTimeoutSpec extends AkkaSpec {
...
+
+      timeoutActor ! TransparentTick
+      timeoutActor ! TransparentTick
+
+      // not triggered by initialTimeout because it was changed by the ticks
+      probe.expectNoMessage(initialTimeout) //  the last set value
+      // now should happen
+      probe.expectMsgType[ReceiveTimeout]
+      system.stop(timeoutActor)
+    }
```
Comment (patriknw) [inline akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: What Johannes was looking for was a test that failed before we fixed the bug but is now passing when the bug is fixed. This version passes on master as well.

It can also easily fail here by changing to:
```
probe.expectNoMessage(initialTimeout + 50.millis)
```

which is too tight for CI env and it will have spurious failures.

Here is a version that fails on master and passes here and verifies that the new timeout is set via the TransparentTick.

```scala
    "work correctly if the timeout is changed multiple times while handling a NotInfluenceReceiveTimeout message" taggedAs TimingTest in {
      val probe = TestProbe()
      val initialTimeout = 500.millis

      val timeoutActor = system.actorOf(Props(new Actor {
        var count = 0
        context.setReceiveTimeout(initialTimeout)

        def receive: Receive = {
          case TransparentTick =>
            context.setReceiveTimeout(initialTimeout * 2)
            count += 1 // do some work then
            context.setReceiveTimeout(initialTimeout)

          case ReceiveTimeout => probe.ref ! ReceiveTimeout
        }
      }))

      probe.expectNoMessage(initialTimeout / 2)
      
      timeoutActor ! TransparentTick
      timeoutActor ! TransparentTick

      // not triggered by initialTimeout because it was changed by the ticks
      // wait shorter than last set value, but longer than initialTimeout
      probe.expectNoMessage(initialTimeout / 2 + 50.millis)
      // now should happen
      probe.expectMsgType[ReceiveTimeout]
      system.stop(timeoutActor)
    }
```

Comment (helena) [reply akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala]: Thanks @patriknw that makes sense now. Pushed.

**Review by patriknw — APPROVED** [review]
LGTM

**Review by jrudolph — APPROVED** [review]
LGTM :+1:

Should we also update the scaladoc on `setReceiveTimeout` to be explicit about that calling `setReceiveTimeout` resets the timeout also during handling a `NotInfluenceReceiveTimeout` message?

**Comment by patriknw** [comment]
> Should we also update the scaladoc on setReceiveTimeout to be explicit about that calling setReceiveTimeout resets the timeout also during handling a NotInfluenceReceiveTimeout message?

I don't think that is needed, it's expected behavior in my opinion

**Comment by patriknw** [comment]
This doesn't cherry pick cleanly to release-2.5 so I think we'll skip the backport. `NotInfluenceReceiveTimeout` messages combined with changing the timeout should be a rare thing.

**Comment by helena** [comment]
Ah, right. Thanks.
